flight/Modules/GPS/GPS.c

```diff
@@ -261,20 +260,22 @@ static void gpsTask(void *parameters)
 			}
 
 			if (res == PARSER_COMPLETE) {
-				timeNowMs = PIOS_Thread_Systime();
-				timeOfLastUpdateMs = timeNowMs;
-				timeOfLastCommandMs = timeNowMs;
+				timeOfLastUpdateMs = PIOS_Thread_Systime();
 			}
 		}
 
 		// Check for GPS timeout
-		timeNowMs = PIOS_Thread_Systime();
+		uint32_t timeNowMs = PIOS_Thread_Systime();
 		if ((timeNowMs - timeOfLastUpdateMs) >= GPS_TIMEOUT_MS) {
 			// we have not received any valid GPS sentences for a while.
 			// either the GPS is not plugged in or a hardware problem or the GPS has locked up.
 			uint8_t status = GPSPOSITION_STATUS_NOGPS;
 			GPSPositionStatusSet(&status);
 			AlarmsSet(SYSTEMALARMS_ALARM_GPS, SYSTEMALARMS_ALARM_ERROR);
+			/* Don't reinitialize too often. */
+			if ((timeNowMs - timeOfLastCommandMs) >= GPS_TIMEOUT_MS) {
+				goto reinit;
```

Comment:
I know goto isn't considered an absolute prohibition anymore, but in this case it does make the code structure fairly hard to read. Would you mind doing this?

 1. move the initialization code into it's own function (this will be useful anyway when you start reparameterizing it)
 2. call that at startup and also here
 3. update the timeOfLastCommandMs / timeOfLastReinitMs here

That should make it quite a bit more readable